// node-libpng/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
/* istanbul ignore file */
var decode_1 = require("./decode");
Object.defineProperty(exports, "readPngFile", { enumerable: true, get: function () { return decode_1.readPngFile; } });
Object.defineProperty(exports, "readPngFileSync", { enumerable: true, get: function () { return decode_1.readPngFileSync; } });
Object.defineProperty(exports, "decode", { enumerable: true, get: function () { return decode_1.decode; } });
var encode_1 = require("./encode");
Object.defineProperty(exports, "writePngFile", { enumerable: true, get: function () { return encode_1.writePngFile; } });
Object.defineProperty(exports, "writePngFileSync", { enumerable: true, get: function () { return encode_1.writePngFileSync; } });
Object.defineProperty(exports, "encode", { enumerable: true, get: function () { return encode_1.encode; } });
var png_image_1 = require("./png-image");
Object.defineProperty(exports, "PngImage", { enumerable: true, get: function () { return png_image_1.PngImage; } });
var is_png_1 = require("./is-png");
Object.defineProperty(exports, "isPng", { enumerable: true, get: function () { return is_png_1.isPng; } });
__exportStar(require("./colors"), exports);
__exportStar(require("./rect"), exports);
__exportStar(require("./xy"), exports);
